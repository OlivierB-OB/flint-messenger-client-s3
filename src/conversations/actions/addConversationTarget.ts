import { IAddConversationTargetAction, ADD_CONVERSATION_TARGET } from '../types';

export function addConversationTarget(conversationId: string, target: string): IAddConversationTargetAction {
  return {
    type: ADD_CONVERSATION_TARGET,
    conversationId,
    target,
  };
}
