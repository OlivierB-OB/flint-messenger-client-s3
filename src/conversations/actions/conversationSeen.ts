import { IConversationSeenAction, CONVERSATION_SEEN } from '../types';

export function conversationSeen(id: string, seenDate: string): IConversationSeenAction {
  return {
    type: CONVERSATION_SEEN,
    id,
    seenDate,
  };
}
