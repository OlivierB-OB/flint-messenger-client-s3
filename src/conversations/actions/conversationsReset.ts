import { IConversationsResetAction, CONVERSATIONS_RESET } from '../types';

export function conversationsReset(): IConversationsResetAction {
  return { type: CONVERSATIONS_RESET };
}
