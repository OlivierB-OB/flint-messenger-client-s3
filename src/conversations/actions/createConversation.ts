import { ICreateConversationAction, CREATE_CONVERSATION } from '../types';

export function createConversation(conversationId: string, targets: string[]): ICreateConversationAction {
  return {
    type: CREATE_CONVERSATION,
    conversationId,
    targets,
  };
}
