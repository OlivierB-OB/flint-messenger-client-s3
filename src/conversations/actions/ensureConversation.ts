import { IEnsureConversationAction, ENSURE_CONVERSATION } from '../types';

export function ensureConversation(
  conversationId: string,
  targets: string[],
  createdAt: string,
): IEnsureConversationAction {
  return {
    type: ENSURE_CONVERSATION,
    conversationId,
    targets,
    createdAt,
  };
}
