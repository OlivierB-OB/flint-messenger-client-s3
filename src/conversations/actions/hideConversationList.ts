import { IToggleConversationListAction, TOGGLE_CONVERSATION_LIST } from '../types';

export function hideConversationList(): IToggleConversationListAction {
  return {
    type: TOGGLE_CONVERSATION_LIST,
    value: false,
  };
}
