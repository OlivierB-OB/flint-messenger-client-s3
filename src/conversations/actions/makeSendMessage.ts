import { batch } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { updateConversation } from './updateConversation';
import { updateMessageEdition } from './updateMessageEdition';
import { updateConversationStatus } from './updateConversationStatus';

const SLEEP = (delay: number) => new Promise((resolve) => setTimeout(resolve, delay));

export function makeSendMessage(conversationId: string) {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    const { conversations, identity } = getState();
    const { messageEdition } = conversations;

    if (!messageEdition) return; // abort

    const createdAt = new Date().toISOString();

    batch(() => {
      dispatch(updateMessageEdition(''));
      dispatch(updateConversationStatus('sending'));
    });

    const { info } = identity;
    if (!info) return dispatch(updateConversationStatus('error'));

    // FIXME HTTP call goes here
    await SLEEP(500);

    batch(() => {
      dispatch(
        updateConversation({
          conversationId,
          createdAt,
          emitter: info.uid,
          content: messageEdition,
        }),
      );
      dispatch(updateConversationStatus('ready'));
    });
  };
}
