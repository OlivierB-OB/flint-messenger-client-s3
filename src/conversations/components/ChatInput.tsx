import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';
import Send from '@material-ui/icons/Send';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { makeSendMessage } from '../actions/makeSendMessage';
import { updateMessageEdition } from '../actions/updateMessageEdition';

interface IChatInputDisplayProps {
  conversationId: string;
  messageEdition: string;
  updateMessageEdition: (text: string) => void;
  sendMessage: (conversationId: string) => void;
}

function ChatInputDisplay({
  messageEdition,
  conversationId,
  updateMessageEdition,
  sendMessage,
}: IChatInputDisplayProps) {
  useEffect(() => updateMessageEdition(''), [updateMessageEdition, conversationId]);
  return (
    <Grid container justify="center">
      <Grid item xs={11} justify="center">
        <TextField
          fullWidth={true}
          value={messageEdition}
          variant="filled"
          onChange={(event) => updateMessageEdition(event.target.value)}
          onKeyDown={(event) => event.key === 'Enter' && sendMessage(conversationId)}
        />
      </Grid>
      <Grid item container xs={1} justify="center" alignItems="center">
        <IconButton aria-label="contacts" onClick={() => sendMessage(conversationId)}>
          <Send fontSize="large" color="primary" />
        </IconButton>
      </Grid>
    </Grid>
  );
}

export interface IChatInputProps {
  conversationId: string;
}

const mapStateToProps = ({ conversations }: IAppState, { conversationId }: IChatInputProps) => ({
  conversationId,
  messageEdition: conversations.messageEdition,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  updateMessageEdition: (text: string) => void dispatch(updateMessageEdition(text)),
  sendMessage: (conversationId: string) => void dispatch(makeSendMessage(conversationId)),
});

export const ChatInput = connect(mapStateToProps, mapDispatchToProps)(ChatInputDisplay);
