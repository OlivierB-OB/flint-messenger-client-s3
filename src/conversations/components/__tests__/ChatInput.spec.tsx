import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { ChatInput } from '../ChatInput';
import { makeSendMessage } from '../../actions/makeSendMessage';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('ChatInput', () => {
  let Connected: any = ChatInput;
  let spySend: jest.Mock;
  let restoreSend: () => void;

  beforeEach(() => {
    [spySend, restoreSend] = makeSendMessage.mock(jest.fn());
  });

  afterEach(() => {
    restoreSend();
  });

  it('should provide messageEdition, conversationId and callbacks to ChatInputDisplay', async () => {
    const store = mockStore({
      conversations: {
        messageEdition: 'the messageEdition',
      },
    });
    const component = shallow(<Connected store={store} conversationId={'the conversationId'} />).dive();
    expect(component).toMatchSnapshot();

    (component.prop('updateMessageEdition') as any)('new value');
    expect(store.getActions()).toEqual([
      {
        messageEdition: 'new value',
        type: 'UPDATE_MESSAGE_EDITION',
      },
    ]);
    store.clearActions();

    (component.prop('sendMessage') as any)(component.prop('conversationId'));
    expect(spySend).toHaveBeenCalledWith('the conversationId');
  });
});
