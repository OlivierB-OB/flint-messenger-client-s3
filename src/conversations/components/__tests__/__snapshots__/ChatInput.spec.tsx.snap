// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInput should provide messageEdition, conversationId and callbacks to ChatInputDisplay 1`] = `
<ChatInputDisplay
  conversationId="the conversationId"
  messageEdition="the messageEdition"
  sendMessage={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  updateMessageEdition={[Function]}
/>
`;
